### DISCUSSION TITLE
Is it possible to make an "implicit join" on this relationship?

### DISCUSSION THREAD

**Discussion by musbur** [sqlalchemy/sqlalchemy#9699] [Usage Questions] [answered]
I have an ORM mapped table:

    class Sequence(Base):
        __tablename__ = "sequence"
        id = Column(Integer, primary_key=True)
        recipe_name = Column(String(100))

In reality, this table of course has a few more columns, but the bulk of the data is in the recipe_name, which is 10-20 characters long. Over time, this table has grown several GB in size. Since there are only a few dozen different recipe names, I want to make this more efficient by storing the recipes in a different table "Recipe" and just referencing them from Sequence with a foreign key. Simple enough, but I don't want to have to change the rest of my code accessing the recipe_name property on Sequence instances. Moreover, since this is a long running process which does thousands operations on Sequence per second I want to have the recipe names accessible in an efficient registry type fashion. So I came up with the solution below (fully self contained and runnable), which works fine except for one thing: I would like to use recipe_name in expressions like

    query(Sequence).filter_by(recipe_name="abc")

without any explicit join() to Recipe. Is this possible at all?

    from sqlalchemy import (create_engine, Column, Integer,
                            String, ForeignKey)
    from sqlalchemy.orm import (sessionmaker, declarative_base,
                                relationship)
    from sqlalchemy.ext.hybrid import hybrid_property

    Base = declarative_base()

    class RecipeRegistry():
        def __init__(self, db):
            self.db = db
            self.data = dict()

        def get(self, name):
            recipe = self.data.get(name)
            if not recipe:
                recipe = (self.db.query(Recipe)
                          .filter_by(name=name)
                          .one_or_none())
                if not recipe:
                    recipe = Recipe(name=name)
                self.data[name] = recipe
            return recipe

    class Recipe(Base):
        __tablename__ = "recipe"
        id            = Column(Integer, primary_key=True)
        name          = Column(String(200), unique=True)

    class Sequence(Base):
        __tablename__ = "sequence"
        id            = Column(Integer, primary_key=True)
        recipe_id     = Column(ForeignKey("recipe.id"))
        recipe = relationship(Recipe)

        @hybrid_property
        def recipe_name(self):
            if not self._recipe:
                raise RuntimeError("Can't happen")
            return self._recipe.name

        @recipe_name.setter
        def recipe_name(self, name):
            self.recipe = _recipe_registry.get(name)

        @recipe_name.expression
        def recipe_name(cls):
            return Recipe.name

    engine = create_engine("sqlite:///test.db")
    Base.metadata.create_all(engine)

    db = sessionmaker(engine)()

    _recipe_registry = RecipeRegistry(db)

    for n in ('Meep', 'Maap', 'Moop', 'Meep'):
        r = Sequence(recipe_name=n)
        db.add(r)
    db.commit()

    rs = (db.query(Sequence)
          .join(Recipe) # I want to get rid of this join
          .filter(Sequence.recipe_name == "Meep"))
    for r in rs:
        print(r)

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think association proxy allows you to do that: https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#module-sqlalchemy.ext.associationproxy

**Reply by musbur** [reply]
I had pored over association_proxy but the examples given dealt with collections over association tables, not anything as simple as my issue. Anyway, replacing the whole hybrid_property shebang with association_proxy worked right out of the box:

    from sqlalchemy import create_engine, Column, Integer, String, ForeignKey
    from sqlalchemy.orm import sessionmaker, declarative_base, relationship
    from sqlalchemy.ext.hybrid import hybrid_property
    from sqlalchemy.ext.associationproxy import association_proxy

    Base = declarative_base()

    class RecipeRegistry():
        def __init__(self, db):
            self.db = db
            self.data = dict()

        def get(self, name):
            recipe = self.data.get(name)
            if not recipe:
                recipe = (self.db.query(Recipe)
                          .filter_by(name=name)
                          .one_or_none())
                if not recipe:
                    recipe = Recipe(name=name)
                self.data[name] = recipe
            return recipe

    class Recipe(Base):
        __tablename__ = "recipe"
        id            = Column(Integer, primary_key=True)
        name          = Column(String(200), unique=True)

    def create_recipe(name):
        return _recipe_registry.get(name)

    class Sequence(Base):
        __tablename__ = "sequence"
        id            = Column(Integer, primary_key=True)
        recipe_id     = Column(ForeignKey("recipe.id"))
        recipe = relationship(Recipe)

        recipe_name = association_proxy("recipe", "name",
                                        creator=create_recipe)

    engine = create_engine("sqlite:///test.db")
    Base.metadata.create_all(engine)

    db = sessionmaker(engine)()

    _recipe_registry = RecipeRegistry(db)

    for n in ('Meep', 'Maap', 'Moop', 'Meep'):
        r = Sequence(recipe_name=n)
        db.add(r)
    db.commit()

    rs = (db.query(Sequence).filter_by(recipe_name="Meep"))
    for r in rs:
        print(r)

**Reply by CaselIT** [reply]
Would you like to suggest an improvement for the association proxy docs so indicate that it can be used also for scalar relationships?

**Reply by zzzeek** [reply]
sure the docs say "simplfying scalar collections" which could mislead one into thinking AP only does collections.  There is "Cascading Scalar Deletes" which illustrates a scalar use but isn't generalized to "AP works for non-collections too"

**Reply by CaselIT** [reply]
I've added a scalar example to the docs here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4592
